src/lib/comms/sol-lwm2m.c

```diff
@@ -1127,6 +1316,124 @@ sol_lwm2m_server_del_registration_monitor(struct sol_lwm2m_server *server,
     return sol_monitors_del(&server->registration, i);
 }
 
+SOL_API struct sol_lwm2m_bootstrap_server *
+sol_lwm2m_bootstrap_server_new(uint16_t port, const char **known_clients)
+{
+    struct sol_lwm2m_bootstrap_server *server;
+    struct sol_network_link_addr servaddr = { .family = SOL_NETWORK_FAMILY_INET6,
+                                              .port = port };
+    int r;
+
+    SOL_LOG_INTERNAL_INIT_ONCE;
+
+    server = calloc(1, sizeof(struct sol_lwm2m_bootstrap_server));
+    SOL_NULL_CHECK(server, NULL);
+
+    server->coap = sol_coap_server_new(&servaddr);
+    SOL_NULL_CHECK_GOTO(server->coap, err_coap);
+
+    server->known_clients = known_clients;
+
+    sol_monitors_init(&server->bootstrap, NULL);
+
+    r = sol_coap_server_register_resource(server->coap,
+        &bootstrap_request_interface, server);
+    SOL_INT_CHECK_GOTO(r, < 0, err_register);
+
+    return server;
+
+err_register:
+    sol_coap_server_unref(server->coap);
+err_coap:
+    free(server);
+    return NULL;
+}
+
+SOL_API void
+sol_lwm2m_bootstrap_server_del(struct sol_lwm2m_bootstrap_server *server)
+{
+    SOL_NULL_CHECK(server);
+
+    sol_coap_server_unref(server->coap);
+
+    sol_monitors_clear(&server->bootstrap);
+    free(server);
+}
+
+SOL_API int
+sol_lwm2m_bootstrap_server_add_request_monitor(struct sol_lwm2m_bootstrap_server *server,
+    void (*cb)(void *data,
+    struct sol_lwm2m_bootstrap_server *server,
+    struct sol_lwm2m_bootstrap_client_info *bs_cinfo),
+    const void *data)
+{
+    struct sol_monitors_entry *m;
+
+    SOL_NULL_CHECK(cb, -EINVAL);
+    SOL_NULL_CHECK(server, -EINVAL);
+
+    m = sol_monitors_append(&server->bootstrap,
+        (sol_monitors_cb_t)cb, data);
+    SOL_NULL_CHECK(m, -ENOMEM);
+    return 0;
+}
+
+SOL_API int
+sol_lwm2m_bootstrap_server_del_request_monitor(struct sol_lwm2m_bootstrap_server *server,
+    void (*cb)(void *data,
+    struct sol_lwm2m_bootstrap_server *server,
+    struct sol_lwm2m_bootstrap_client_info *bs_cinfo),
+    const void *data)
+{
+    int i;
+
+    SOL_NULL_CHECK(server, -EINVAL);
+    SOL_NULL_CHECK(cb, -EINVAL);
+
+    i = sol_monitors_find(&server->bootstrap, (sol_monitors_cb_t)cb, data);
+    if (i < 0)
+        return i;
+
+    return sol_monitors_del(&server->bootstrap, i);
+}
+
+SOL_API int
+sol_lwm2m_client_add_bootstrap_finish_monitor(struct sol_lwm2m_client *client,
+    void (*cb)(void *data,
+    struct sol_lwm2m_client *client,
+    enum sol_lwm2m_bootstrap_event event),
+    const void *data)
+{
+    struct sol_monitors_entry *m;
+
+    SOL_NULL_CHECK(cb, -EINVAL);
+    SOL_NULL_CHECK(client, -EINVAL);
+
+    m = sol_monitors_append(&client->bootstrap,
+        (sol_monitors_cb_t)cb, data);
+    SOL_NULL_CHECK(m, -ENOMEM);
+    return 0;
+}
+
+SOL_API int
+sol_lwm2m_client_del_bootstrap_finish_monitor(struct sol_lwm2m_client *client,
```

Comment:
the functions `sol_lwm2m_client_del_bootstrap_finish_monitor()`, `sol_lwm2m_client_add_bootstrap_finish_monitor()`, `sol_lwm2m_bootstrap_server_del_request_monitor()` and 
`sol_lwm2m_bootstrap_server_add_request_monitor()` are basically the same thing.

You should do something like:
```C
static int
add_to_monitors(struct sol_monitors *monitors, sol_monitors_cb_t cb, const void *data)
{
   struct sol_monitors_entry *m;
   SOL_NULL_CHECK(cb, -EINVAL);

   m = sol_monitors_append(monitors, cb, data);
   SOL_NULL_CHECK(m, -ENOMEM);
   return 0;
}
static int
remove_from_monitors(struct sol_monitors *monitors, sol_monitors_cb_t cb, const void *data)
{
   int i;
   SOL_NULL_CHECK(cb, -EINVAL);
   i = sol_monitors_find(monitors, cb, data);
   SOL_INT_CHECK(i, < 0, i);
   return sol_monitors_del(monitors, i);
}

....
sol_lwm2m_client_del_bootstrap_finish_monitor(...)
{
   ...
   return remove_from_monitors(&client->bootstrap, cb, data);
}

//Do the same thing for ADD
//If possible try to use these new functions to the old monitors add/del that are already present in the LWM2M code
```